pkg/tagset/tags.go

```diff
@@ -0,0 +1,145 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.Datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package tagset
+
+import (
+	"encoding/json"
+	"sort"
+	"strings"
+
+	"github.com/twmb/murmur3"
+)
+
+// Tags contains a set of tags, along with a 64-bit hash considered to be
+// unique to that set of tags. Tags in the tagset are unique. The order of the
+// tags is undefined.
+//
+// Tags are always handled by pointer (`*tagset.Tags`).  Avoid using the nil
+// form of this pointer; in cases where no tags are necessary, use
+// `tagset.EmptyTags` instead.  This avoids many unnecessary nil checks.
+//
+// The constructor functions associated with this type use the default factory.
+type Tags struct {
+	// Tags is the central struct in this package, so instances are created
+	// directly in many locations in the package. All must adhere to the
+	// invariants described here.
+
+	// tags are the tags contained in this tagset. They must be unique.
+	tags []string
+
+	// hashes are the hashes of the tags at the same index in the tags field
+	hashes []uint64
+
+	// hash is the xor of all values in the hashes field
+	hash uint64
+}
+
+// calcHashes calculates the per-tag hashes and the overall hash for the
+// given slice of tags.
+func calcHashes(tags []string) (hashes []uint64, hash uint64) {
+	hashes = make([]uint64, len(tags))
+	for i, t := range tags {
+		h := murmur3.StringSum64(t)
+		hashes[i] = h
+		hash ^= h
+	}
+	return
+}
+
+// String returns a human-readable form of this tagset.
+func (tags *Tags) String() string {
+	return strings.Join(tags.tags, ", ")
+}
+
+// Len returns the number of tags in the tagset
+func (tags *Tags) Len() int {
+	return len(tags.tags)
+}
+
+// MarshalJSON implements json.Marshaler to marshal the tag set as a JSON
+// array.
+func (tags *Tags) MarshalJSON() ([]byte, error) {
+	return json.Marshal(tags.tags)
+}
+
+// Hash returns the hash of this tagset
+func (tags *Tags) Hash() uint64 {
+	return tags.hash
+}
+
+// Sorted returns a copy of the tags in this tagset, sorted. This is intended
+// for ease of assertions in tests, and not for use in production code.
+func (tags *Tags) Sorted() []string {
```

Comment:
If we use it just in tests, could we make it private?

Reply:
It's used in tests all over the codebase, so that's not possible.